snapshot/driver.go

```diff
@@ -0,0 +1,310 @@
+// The diff will be captured at tmpLocation, as the layer is applied.
+//
+// Running a Container
+//
+// To run a container, we simply provide Manager.Prepare the diffPath
+// of the image we want to start the container from. After mounting, the
+// prepared path can be used directly as the container's filesystem:
+//
+// 	mounts, err := sm.Prepare(containerRootFS, imageDiffPath)
+//
+// The returned mounts can then be passed directly to the container runtime. If
+// one would like to create a new image from the filesystem,
+// Manager.Commit is called:
+//
+// 	if err := sm.Commit(newImageDiff, containerRootFS); err != nil { ... }
+//
+// Alternatively, for most container runs, Manager.Rollback will be
+// called to signal Manager to abandon the changes.
+type Driver interface {
+	// Prepare returns a set of mounts corresponding to an active snapshot
+	// transaction, identifieid by the provided transaction key.
+	//
+	// If a parent is provided, after performing the mounts, the destination
+	// will start with the content of the parent. Changes to the mounted
+	// destination will be captured in relation to the provided parent. The
+	// default parent, "", is an empty directory.
+	//
+	// The changes may be saved to a new snapshot by calling Commit. When one
+	// is done with the transaction, Remove should be called on the key.
+	//
+	// Multiple calls to Prepare or View with the same key should fail.
+	Prepare(key, parent string) ([]containerd.Mount, error)
+
+	// View behaves identically to Prepare except the result may not be committed
+	// back to the snapshot manager. View returns a readonly view on the
+	// parent, with the transaction tracked by the given key.
+	//
+	// This method operates identically to Prepare, except that Mounts returned
+	// may have the readonly flag set. Any modifications to the underlying
+	// filesystem will be ignored.
+	//
+	// Commit may not be called on the provided key. To collect the resources
+	// associated with key, Remove must be called with key as the argument.
+	View(key, parent string) ([]containerd.Mount, error)
+
+	// Commit captures the changes between key and its parent into a snapshot
+	// identified by name.  The name can then be used with the driver's other
+	// methods to create subsequent snapshots.
+	//
+	// A snapshot will be created under name with the parent that started the
+	// transaction.
+	//
+	// Commit may be called multiple times on the same key. Snapshots created
+	// in this manner will all reference the parent used to start the
+	// transaction.
+	Commit(name, key string) error
+
+	// Mounts returns the mounts for the transaction identified by key. Can be
+	// called on an read-write or readonly transaction.
+	//
+	// This can be used to recover mounts after calling View or Prepare.
+	Mounts(key string) ([]containerd.Mount, error)
+
+	// Remove abandons the transaction identified by key. All resources
+	// associated with the key will be removed.
+	Remove(key string) error
```

Comment:
@samuelkarp I agree this isn't ideal. Today, there is some duality between "read write" layers and committed layers and there is a feeble attempt to do this in this proposal. I'll take another pass at to see we can reflect this.